shop/rest/filters.py

```diff
@@ -8,31 +8,34 @@
 from rest_framework.filters import BaseFilterBackend
 
 
-class _CMSPagesFilterBackend(BaseFilterBackend):
+class CMSPagesFilterBackend(BaseFilterBackend):
+    """
+    Use this backend to only show products assigned to the current page.
+    """
+
+    cms_pages_fields = ['cms_pages', ]
+
+    def _get_filtered_queryset(self, current_page, queryset):
+        filter_by_cms_page = (Q((field, current_page)) for field in self.cms_pages_fields)
+        return queryset.filter(reduce(operator.or_, filter_by_cms_page)).distinct()
+
     def filter_queryset(self, request, queryset, view):
-        """
-        Restrict queryset to entities which have one ore more relations to one or more CMS pages.
-        """
         cms_pages_fields = getattr(view, 'cms_pages_fields', self.cms_pages_fields)
         if not isinstance(cms_pages_fields, (list, tuple)):
             msg = "`cms_pages_fields` must be a list or tuple of fields referring to djangoCMS pages."
             raise ImproperlyConfigured(msg)
         current_page = request.current_page
         if current_page.publisher_is_draft:
             current_page = current_page.publisher_public
-        filter_by_cms_page = (Q((field, current_page)) for field in cms_pages_fields)
-        queryset = queryset.filter(reduce(operator.or_, filter_by_cms_page)).distinct()
-        return queryset
+        return self._get_filtered_queryset(current_page, queryset)
 
 
-class CMSPagesFilterBackend(type):
+class RecursiveCMSPagesFilterBackend(CMSPagesFilterBackend):
     """
-    Class builder to filter by CMS pages.
-    A class returned by this backend, shall be added to filter_backends, when CMS pages are
-    used to emulate product categories.
+    Use this backend to show products assigned to the current page or any of its descendants.
     """
-    def __new__(cls, cms_pages_fields=('cms_pages',)):
-        bases = (_CMSPagesFilterBackend,)
-        attrs = {'cms_pages_fields': cms_pages_fields}
-        new_class = type(cls.__name__, bases, attrs)
-        return new_class
+
+    def _get_filtered_queryset(self, current_page, queryset):
+        pages = current_page.get_descendants(include_self=True)
+        filter_by_cms_page = (Q((field + "__in", pages)) for field in self.cms_pages_fields)
+        return queryset.filter(reduce(operator.or_, filter_by_cms_page)).distinct()
```

Comment:
I know, this was already part of the old code, but we should use ``functools.reduce`` instead of just ``reduce``.